File: Leetcode/sort_and_search/56_MergeIntervals/Solution.py
class Solution(object):
    def merge(self, intervals):
        """
        :type intervals: List[List[int]]
        :rtype: List[List[int]]
        """
        if not intervals: return []
        xs = sorted(intervals, key = lambda x: x[0])
        res = []
        def mergeInterval(interval1, interval2):
            res = [min(interval1[0], interval2[0]), max(interval1[1], interval2[1])]
            return res
        res.append(xs[0])
        for i in range(1, len(xs)):
            if xs[i][0] <= res[-1][1]:
                temp = res.pop()
                res.append(mergeInterval(temp, xs[i]))
            else:
                res.append(xs[i])
        return res

File: Leetcode/sort_and_search/56_MergeIntervals/Solution.java
class Solution {
    public int[][] merge(int[][] intervals) {
        if(intervals.length<2)
            return intervals;
        Arrays.sort(intervals, new Comparator<int[]>() {
            public int compare(int[] a, int[] b) {
                return a[0]-b[0];
            }
        });
        List<int[]> list = new LinkedList<>();
        int[] temp = intervals[0];
        for(int i = 1; i<intervals.length; ++i){
            if (temp[1]>=intervals[i][0]){
                int[] temp2 = mergeInterval(temp, intervals[i]);
                temp = temp2;
            }else{
                list.add(temp);
                temp = intervals[i];
            }
            if(i==intervals.length-1)
                list.add(temp);
        }
        int[][] res = new int[list.size()][2];
        for(int i = 0; i<res.length; ++i){
            res[i] = list.get(i);
        }
        return res;
    }
    private int[] mergeInterval(int[] interval1, int[] interval2){
        return new int[] {Math.min(interval1[0], interval2[0]), Math.max(interval1[1], interval2[1])};
    }
}


